src/ExactDiagonalization/basis_fock.jl

```diff
@@ -0,0 +1,181 @@
+"""
+    build_basis(addr::AbstractFockAddress)
+    build_basis(::Type{<:AbstractFockAddress}) -> basis
+
+Return all possible Fock states of a given type as a vector. This method is _much_ faster
+than `build_basis(::AbstractHamiltonian, ...)`, but does not take matrix blocking into
+account. This version of `build_basis` accepts no additional arguments.
+
+All address types except [`OccupationNumberFS`](@ref Main.Rimu.OccupationNumberFS) are
+supported.
+
+Returns a sorted vector of length equal to the [`dimension`](@ref) of `addr`.
+"""
+function build_basis(addr::AbstractFockAddress)
+    return build_basis(typeof(addr))
+end
+
+###
+### BoseFS
+###
+# Algorithm:
+# Given an address type of N particles in M modes,
+# For all n_1 ∈ (0 ... N)
+# - put n_1 particles in the first mode
+# - recursively call the function to put the remaining (N-n_1) particles into (M-1) modes.
+# The recursion is parallelised through spawning tasks.
+
+# this is equivalent to `dimension(BoseFS{n,m})`, but does not return a `BigInt`.
+_bose_dimension(n, m) = binomial(n + m - 1, n)
+
+function build_basis(::Type{<:BoseFS{N,M}}) where {N,M}
+    T = typeof(near_uniform(BoseFS{N,M}))
+    result = Vector{T}(undef, _bose_dimension(N, M))
+    _bose_basis!(result, (), 1, N, Val(M), 2 * Threads.nthreads())
+    return result
+end
+
+# Multithreaded version - attempts to spawn `n_tasks` tasks.
+@inline function _bose_basis!(
+    result::Vector, postfix, index, remaining_n, ::Val{M}, n_tasks::Int
+) where {M}
+    @sync if M < 5 || remaining_n ≤ 1 || n_tasks ≤ 0
+        _bose_basis!(result, postfix, index, remaining_n, Val(M))
+    else
+        n_tasks ÷= 2
+        Threads.@spawn begin
+            _bose_basis!(result, $(0, postfix...), $index, $remaining_n, Val(M-1), $n_tasks)
+        end
+        index += _bose_dimension(remaining_n, M - 1)
+        Threads.@spawn begin
+            _bose_basis!(
+                result, $(1, postfix...), $index, $(remaining_n - 1), Val(M-1), $n_tasks
+            )
+        end
+        index += _bose_dimension(remaining_n - 1, M - 1)
+        for n in 2:remaining_n
+            _bose_basis!(result, (n, postfix...), index, remaining_n - n, Val(M-1))
+            index += _bose_dimension(remaining_n - n, M - 1)
+        end
+    end
+end
+
+@inline function _bose_basis!(
+    result::Vector{T}, postfix, index, remaining_n, ::Val{M}
+) where {M,T}
+    if remaining_n == 0
+        @inbounds result[index] = T((ntuple(Returns(0), Val(M))..., postfix...))
+    elseif remaining_n == 1
+        _basis_basecase_N1!(result, postfix, index, Val(M))
+    elseif M == 1
+        @inbounds result[index] = T((remaining_n, postfix...))
+    elseif M == 2
+        _bose_basis_basecase_M2!(result, postfix, index, remaining_n)
+    elseif M == 3
```

Comment:
Similarly, why is `M=3` a special case?

Reply:
I could add more base cases, but I arbitrarily decided to stop at `M=3`.